### PR TITLE
Fix quicserver binding when duplicate entries exist

### PR DESCRIPTION
**Pull request by nhorman**
In testing the quic demos, I found that the quicserver refused to start for me, indicating an inability to bind a socket to listen on

The problem turned out to be that getaddrinfo on my system was returning multiple entries, due to the fact that /etc/host maps the localhost host name to both ipv4 (127.0.0.1) and ipv6 (::1), but returns the latter as an ipv4 mapped address (specifying family == AF_INET)

It seems like the proper fix would be to modify the /etc/hosts file to not make that mapping, and indeed that works.  However, since several distribution ship with this setup, it seems like it is worthwhile to manage it in the server code.

its also that some other application may be bound to a given address/port leading to failure, which I think could be considered erroneous, as any failure for the full addrinfo list in quicserver would lead to a complete failure

Fix this by modifying the create_dgram_bio function to count the number of sockets is successfully binds/listens on, skipping any failures, and only exit the application if the number of bound sockets is zero.

Fixes openssl/project#253

### CODE REVIEW

File: util/quicserver.c [hunk 1]
```diff
@@ -94,23 +95,22 @@ static BIO *create_dgram_bio(int family, const char *hostname, const char *port)
...
         }
 
         /* Set to non-blocking mode */
         if (!BIO_socket_nbio(sock, 1)) {
             BIO_closesocket(sock);
-            sock = -1;
             continue;
         }
+        bound_socks++;
     }
```
Comment (mattcaswell) [inline util/quicserver.c:107]: This loop doesn't actually quite look right (a pre-existing issue I think). There should be a "break" at the end of this, i.e. if we get to the end of the "for" block and haven't hit "continue", then we have successfully bound a socket and we should stop looking for more sockets. As things stand we loop through all the sockets, create an fd for each one, but then only use the last one.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: util/quicserver.c (hunk 1 above)
Comment (t8m) [reply util/quicserver.c:107]: With `break` `ai == NULL` could be used as the failure condition.

Comment (hlandau) [reply util/quicserver.c:107]: Agreed, this leaks sockets as-is...

File: util/quicserver.c [hunk 2]
```diff
@@ -94,23 +95,23 @@ static BIO *create_dgram_bio(int family, const char *hostname, const char *port)
...
         }
+        bound_socks++;
+        break; /* stop searching if we found an addr */
     }
 
     /* Free the address information resources we allocated earlier */
     BIO_ADDRINFO_free(res);
 
-    /* If sock is -1 then we've been unable to connect to the server */
-    if (sock == -1)
```
Comment (mattcaswell) [inline util/quicserver.c:113]: So, why is this change necessary now? If `bound_socks == 0` then won't `sock` always be `-1` (if you don't remove the `sock = -1` assignments?

File: util/quicserver.c (hunk 1 above)
Comment (nhorman) [reply util/quicserver.c:108]: I just pushed  a fixup commit to do exactly that.  We now break the loop as soon as we successfully bind and listen on a socket, which mirrors what the clients do on connect

**Review by t8m — APPROVED** [review]

**Comment by nhorman** [comment]
force pushing a zero-change update to restart ci.  The quicapi test failed in ci, but passes locally so I'm restarting

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by mspncp** [comment]
typo: binidng -> binding

(commit subject line and pr title)

**Review by mattcaswell — APPROVED** [review]
Approved assuming the nit is fixed

**Comment by nhorman** [comment]
typo fixed up

**Comment by mspncp** [comment]
Typo still persists in subject line of commit message, but this can be fixed by whoever merges the pr.

**Comment by mattcaswell** [comment]
Pushed to master and 3.2
